### DISCUSSION TITLE
Unable to add tags to a new vlan in 'scripts'

### DISCUSSION THREAD

**Discussion by PieterL75** [netbox-community/netbox#6213] [Q&A] [answered]
### NetBox version

v2.11.0

### Python version

3.7

### Steps to Reproduce

Create a new script with a portion to create a new vlan with tags
         vlangroup=VLANGroup.objects.get(id=3)
         newvid = '1234'
         vlanname = "testvlan"
         newvlan = VLAN(
            vid = newvid,
            name = vlanname,
            group = vlangroup,
            tags = [3,9,16]
         )
         newvlan.save()
         self.log_success(f"Vlan tags : {newvlan.tags}")

         newvlan=VLAN.objects.get(id=newvlan.id)
         self.log_success(f"Vlan tags : {newvlan.tags}")

The 3,9,16 are valid IDs for tags.

### Expected Behavior

A new vlan is created, with the 3 tags assigned to the vlan

### Observed Behavior

The vlan is created, but no tags are seen on the vlan.
I tried several methods of adding values to the vlan (single number, array of numbers, array of tag-objects), but I never get the tags on the newvly created vlan
The first log_success shows "[16, 9, 8]"
after getting the vlan again, the log_success shows "ipam.VLAN.None"

**Comment by jeremystretch** [comment]
Tags use a [many-to-many relationship](https://docs.djangoproject.com/en/3.2/topics/db/examples/many_to_many/), which is handled a bit differently from model fields. You'll need to save the new instance first, then call `.add()` or `.set()` to assign the tags.

**Comment by PieterL75** [comment, marked as answer]
Jeremy,

Thank you for taking time to look into this.

I have this code now : 

`
vlangroup=VLANGroup.objects.get(id=3)
newvid = '1234'
vlanname = "testvlan"
newvlan = VLAN(
vid = newvid,
name = vlanname,
group = vlangroup,
)
newvlan.save()
self.log_success(f"Vlan tags : {newvlan.tags}")

newvlan.tag.add(16)
newvlan.tag.add(9)
newvlan.tag.add(3)
newvlan.save()

newvlan=VLAN.objects.get(vid=1234)
self.log_success(f"Vlan tags : {newvlan.tags}")
`

But I still get the 'ipam.VLAN.None'. 

Really strange, it must be my coding skill I guess. Unfortunattely, I'm not able to find some good example

Pieter

**Reply by PieterL75** [reply]
Jeremy,

I checked the "/netbox/extras/tests/test_tags.py" page, and I saw that the tags are added with the tag.name, and not the tag.id !

After I changed the code to use the tag.name, it all works fine !

Much appreciated

Pieter

**Reply by candlerb** [reply]
[nbshell](https://netbox.readthedocs.io/en/stable/administration/netbox-shell/) is a good way to test this interactively:

```
>>> newvlan = VLAN(vid=1234, name='testvlan')
>>> newvlan.save()
>>> newvlan.tags
<taggit.managers._TaggableManager object at 0x7ff9bd1ffa58>
>>> newvlan.tags.all()
<RestrictedQuerySet []>
>>> newvlan.tag.add(16)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
AttributeError: 'VLAN' object has no attribute 'tag'
>>>
```

The first problem is that in your code you said "tag" not "tags".  It should have aborted with the above error.  But let's assume you meant "tags"

```
>>> newvlan.save()
>>> newvlan.tags.add("abc")
>>> newvlan.save()
>>> newvlan.tags
<taggit.managers._TaggableManager object at 0x7ff9bd1ffa58>
>>> newvlan.tags.all()
<RestrictedQuerySet [<Tag: abc>]>
>>> VLAN.objects.get(vid=1234).tags.all()
<RestrictedQuerySet [<Tag: abc>]>
>>>
```

Your other problem is that 'tags' is a manager, not a list.  This allows you to apply additional filtering:

```
>>> VLAN.objects.get(vid=1234).tags.filter(name__in=["abc","xyz"])
<RestrictedQuerySet [<Tag: abc>]>
>>> VLAN.objects.get(vid=1234).tags.filter(name__in=["def","ghi"])
<RestrictedQuerySet []>
```

Or you can call `all()` to get everything:

```
>>> VLAN.objects.get(vid=1234).tags
<taggit.managers._TaggableManager object at 0x7ff9bbcf6d68>
>>> VLAN.objects.get(vid=1234).tags.all()
<RestrictedQuerySet [<Tag: abc>]>
```

It's certainly confusing that the manager renders as "ipam.VLAN.None" when interpolated into a string, but apparently that's how it is:

```
>>> print(f"VLAN tags: {VLAN.objects.get(vid=1234).tags}")
VLAN tags: ipam.VLAN.None
>>> print(f"VLAN tags: {VLAN.objects.get(vid=1234).tags.all()}")
VLAN tags: <RestrictedQuerySet [<Tag: abc>]>

>>> VLAN.objects.get(vid=1234).tags
<taggit.managers._TaggableManager object at 0x7ff9bbcf6d68>
>>> str(VLAN.objects.get(vid=1234).tags)
'ipam.VLAN.None'
```

**Reply by PieterL75** [reply]
awesome explanation Brian !
I'll give that nbshell a go, it will save me time to learn how this all works :-)

Pieter
